source/store.ts

```diff
@@ -0,0 +1,190 @@
+		this.expiration = options.windowMs / 1000 // [in seconds]
+	}
+
+	/**
+	 * Method to prefix the keys with the given text.
+	 *
+	 * @param key {string} - The key.
+	 *
+	 * @returns {string} - The text + the key.
+	 */
+	prefixKey(key: string): string {
+		return `${this.prefix}${key}`
+	}
+
+	/**
+	 * Method to increment a client's hit counter.
+	 *
+	 * @param key {string} - The identifier for a client.
+	 *
+	 * @returns {IncrementResponse} - The number of hits and reset time for that client.
+	 */
+	async increment(key: string): Promise<IncrementResponse> {
+		const response = (await this.updateKey(
+			this.prefixKey(key),
+			+1,
+		)) as IncrementResponse // We can cast this for sure because we're passing delta as +1
+		return response
+	}
+
+	/**
+	 * Method to decrement a client's hit counter.
+	 *
+	 * @param key {string} - The identifier for a client
+	 */
+	async decrement(key: string): Promise<void> {
+		await this.updateKey(this.prefixKey(key), -1)
+	}
+
+	/**
+	 * Method to reset a client's hit counter.
+	 *
+	 * @param key {string} - The identifier for a client.
+	 */
+	async resetKey(key: string): Promise<void> {
+		const prefixedKey = this.prefixKey(key)
+
+		// Delete the record for that key.
+		await new Promise<void>((resolve, reject) => {
+			this.client.del(prefixedKey, (error) => {
+				if (error) {
+					reject(error)
+					return
+				}
+
+				resolve()
+			})
+		})
+	}
+
+	private async updateKey(
+		key: string,
+		delta: number,
+	): Promise<IncrementResponse | void> {
+		// Find the existing record for that key.
+		let record = await new Promise<ClientRecord>((resolve, reject) => {
+			this.client.get(key, (error, data) => {
+				if (error) {
+					reject(error)
+					return
+				}
+
+				resolve(data as ClientRecord)
+			})
+		})
+
+		if (record === undefined && delta < 0) {
+			// If the record does not exist, and we are supposed to decrement the key,
+			// we don't need to do anything, so we return.
+			return
+		}
+
+		if (record === undefined) {
+			// If a record does not exist, and must be incremented, then add a record
+			// for that key.
+			record = await new Promise<ClientRecord>((resolve, reject) => {
+				const payload: ClientRecord = { hits: 1, time: Date.now() }
+
+				this.client.add(key, payload, this.expiration, (error) => {
+					if (error) {
+						reject(error)
+						return
+					}
+
+					resolve(payload)
+				})
+			})
+		} else {
+			// If it does exist, simply change the hit count by `delta`.
+			record = await new Promise<ClientRecord>((resolve, reject) => {
+				const payload: ClientRecord = { ...record, hits: record.hits + delta }
+				const expiration = Math.floor(
+					// = window - (time elapsed since start of window) [in seconds]
+					this.expiration - (Date.now() - payload.time) / 1000,
+				)
+
+				this.client.replace(key, payload, expiration, (error) => {
+					if (error) {
+						reject(error)
+						return
+					}
+
+					resolve(payload)
+				})
+			})
+		}
```

Comment:
I liked the old flow of calling incr/decr. Calling get then create/update leaves it you open to race conditions where a bunch of requests come in at once, and all get the initial value then all overwrite eachother's additions.

IMO, removing this function and just letting increment and decrement each have their own logic probably makes more sense here.

Reply:
We could do that, but then we won't be able to return an accurate `resetTime` to the middleware.

Reply:
Oh, that's a good point. I should familiarize myself with memcahed's commands.

Reply:
What about `incr` followed by either creating the key if it's not found, or `mg` (meta get), with the `t` to return the expiration time in seconds if the `incr` worked.

It's still two commands, but there's no risk of collision (after the initial creation) this way.

Reply:
Oooh that sounds good, let me do that. Thanks!

Reply:
Maybe we should switch to https://github.com/electrode-io/memcache/tree/master/packages/memcache-client? It has support for sending raw commands, and is more actively maintained.

Reply:
Oh, I like that it returns promises also, so we can await it. Yeah, that seems like a good idea.

Reply:
It looks like you can use `memcached.command()` in the old memcached library to send arbitrary commands. It's documented as private, but it is documented.

It might be worth trying to build a compatibility layer to let existing installs continue to use the old memcache library, even if we don't recommend it for new installations. 

Somewhat similar to our legacy store wrapper in express rate limit.

Reply:
oh, snap,  the [Meta Arithmetic](https://github.com/memcached/memcached/blob/master/doc/protocol.txt#L851) command, `ma` is exactly what we want.

`ma mykey N60 v t` will:
 * create mykey if it doesn't exist with a expiration time of 60 seconds (`N60`)
 * increment it (`MI`, which is the default so I left it out) by 1 (`D1`, also default)
 * return the new value (`v`) and the expiration time (`t`)

That's a single memcached request per hit, regardless of weather or not the key already existed.

Similarly, `ma mykey N60 MD v t` will decrement (`MD`) it instead of incrementing.

I'm going to see if I can find a memcached client that supports meta commands, and I might send some PRs if I can't find one.

Reply:
I just filed a ticket on memcached-client requesting support for meta commands. I'm kind of afraid that I just signed myself up for a ton of work, but I think it will make this a better library in the long run.

https://github.com/electrode-io/memcache/issues/22

Reply:
> oh, snap, the [Meta Arithmetic](https://github.com/memcached/memcached/blob/master/doc/protocol.txt#L851) command, ma is exactly what we want.

It is!! Great discovery!

> https://github.com/electrode-io/memcache/issues/22

Thanks for opening the issue, I'll do my best to help implement it.

Reply:
I don't see any reply on the issue.. should we try pinging the maintainers?

Otherwise we could try using the private `.command` method on the library we are using right now.

Reply:
Yeah, might as well try that